### ISSUE TITLE
pytorch fp32的训练

### ISSUE THREAD

**Issue by ccssu** [Oneflow-Inc/one-yolov5#37]
### 背景： 
我们这里最新的结果（即 https://github.com/Oneflow-Inc/one-yolov5/issues/26#issuecomment-1251671595) 还是达不到torch的精度，
其中metrics/mAP_0.5:0.95差了**1.468**,目前eval，inference，export等功能都是正常的，并且模型初始化一致的情况下多卡loss是可以完全对齐的。在loss对齐的情况下做了精度验证和pytorch一样，就是训练到不了官方给的精度。下面将进行一次pytorch fp32的训练。
###  pytorch fp32的训练
机器：oneflow27-root

配置项:
1 . shuffle  = True
2.  sync-bn = False
3.  amp = False  # 使用FP32训练
4. --workers 8
 
目前数据

目前opochs:  300 (*全部数据可见 [result.csv](https://paste.ubuntu.com/p/m3zs8M686v/)*)
 | |metrics/mAP_0.5|metrics/mAP_0.5:0.95|
 | --- | --- | --- |
 |当前训练值| 43.350 | 26.233 |
| 目标值 |  45.6 | 27.7 | 
|当前训练值 - 目标值|-2.250 | -1.467|

下图是根据目前数据画出的map0.5 和 map0.5-0.95 趋势图

![image](https://user-images.githubusercontent.com/109639975/192171375-e330b50e-9f8a-4828-a217-a346594da869.png)

**Comment by ccssu** [comment]
## 两次实验结果对比
- 这里当前训练代表：https://github.com/Oneflow-Inc/one-yolov5/issues/37#issue-1384553187 的训练结果(即上面这次实验)
- 这里上次训练代表：https://github.com/Oneflow-Inc/one-yolov5/issues/26#issuecomment-1251671595 的训练结果 
###  $mAP$值
 | |metrics/mAP_0.5|metrics/mAP_0.5:0.95|
 | --- | --- | --- |
 |当前训练值| 43.350 | 26.233 |
|上次训练值| 43.524 | 26.232 |
|当前训练值-上次训练值|  -0.174 | 0.001 |

这里可以看到两次实验的 metrics/mAP_0.5:0.95的值几乎没有差距。

**Comment by yuanms2** [comment]
oneflow 版本的yolo 和 pytorch 版本的yolo 代码长得完全一样吗，是修改import 就可以切换的吗？

**Comment by ccssu** [comment]
> oneflow 版本的yolo 和 pytorch 版本的yolo 代码长得完全一样吗，是修改import 就可以切换的吗？
只能说是大体上长得基本一致，目前这份yolo是通过pytorch版本(*版本号48a85314bc80d8023c99bfb114cea98d71dd0591*)改的,单单只修改import好像不行，（比如：pytorch 版本中train.py 中 dist.init_process_group(backend="nccl" if dist.is_nccl_available() else "gloo") 这一行得改）。

**Comment by yuanms2** [comment]
我们能不能尽力做到只修改import 就可以切换，这样更容易定位二者的差异

**Comment by ccssu** [comment]
> > 配置项:
> > 1 . shuffle = True
> > 2. sync-bn = False
> > 3. amp = False # 使用FP32训练
> > 4. --workers 8
> 
> 为什么训Pytorch fp32 要把 sync-bn 关掉呢

按照了[torch官网](https://github.com/ultralytics/yolov5#pretrained-checkpoints)给的训练方式(sync-bn默认是关掉的) 这次训练就只改为了使用FP32训练 。

![image](https://user-images.githubusercontent.com/109639975/192177761-af8bc914-f5c2-4f99-af3d-51743648a776.png)

**Comment by BBuf** [comment]
> 我们能不能尽力做到只修改import 就可以切换，这样更容易定位二者的差异

对于yolov5，几乎就是只改了import。

**Comment by BBuf** [comment]
> 

实验结论：在fp32数据类型下，oneflow后端的yolov5相比于pytorch的yolov5在coco数据集上的map_0.5（oneflow: 43.35 vs pytorch: 43.524）和map_0.5:0.95(oneflow: 26.233 vs pytorch: 26.232)基本一致，可以认为fp32模式训练的目标精度已经对齐了。

**Comment by BBuf** [comment]
> 我们能不能尽力做到只修改import 就可以切换，这样更容易定位二者的差异

过程中暴露了很多api没有对齐以及算子方面的bug，这些bug目前在master全都解决掉了，所以现在基本可以做到只改import就切换。
